[BE-65] feat: 프로필 조회, 업데이트 기능 구현

xxbeann opened the pull request:
## #️⃣연관된 이슈

> #156 

## 📝작업 내용

> 사용자의 프로필을 조회하고 업데이트하는 기능입니다.


### 스크린샷 (선택)
<img width="1416" height="584" alt="image" src="https://github.com/user-attachments/assets/4f60f019-96a3-4f55-ae30-28bdc559c36c" />

## 💬리뷰 요구사항(선택)

> PhysicalInfo 클래스 내에서 DB 저장전 검증 로직이 필요할까요? 


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **신기능**
  * 프로필 조회 API 추가: 내 프로필 정보를 확인할 수 있습니다.
  * 프로필 수정 API 추가: 이름, 이메일, 닉네임, 전화번호, 키, 몸무게, 혈액형, 기타 메모(etc) 수정 가능.
  * 닉네임 중복 확인 API 재도입: 사용 가능 여부를 확인할 수 있습니다.

* **검증 강화 / 버그 수정**
  * 입력 검증 확대 및 오류 메시지 보강: 항목별 형식·최대길이 적용, 기타 메모 200자 제한 및 관련 안내 메시지 추가.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

xxbeann committed:
feat: User엔티티 etc(기타 사항)추가

xxbeann committed:
chore: 컨트롤러 순서 재배치

xxbeann committed:
feat: 사용자 프로필 조회 DTO 작성

xxbeann committed:
feat: 프로필 조회 Controller 작성

# Conflicts:
#	src/main/java/com/econo_4factorial/newproject/user/controller/UserController.java

xxbeann committed:
feat: 사용자 프로필 조회기능 구현

xxbeann committed:
feat: 프로필 업데이트 메서드 추가 및 UserInfo 값 업데이트 함수 리팩토링

User 도메인에서 사용자 프로필을 업데이트 할 수 있는 메서드를 추가했습니다.
또한, 기존 UserInfo에 존재하던 updateBasicInformation 메서드를 제거하고,
각 필드를 업데이트 할 수 있는 updateEmail, updateName, updatePhoneNumber를 추가했습니다.

기존 updateBasicInformation은 메서드명과 달리 email과 phoneNumber의 값만 업데이트하는 동작이었습니다.
그러던 도중 updateUserProfile 메서드에서 UserInfo의 모든 필드를 업데이트 하는 로직이 필요했습니다.
2개의 값을 업데이트하는 메서드, 3개의 값을 한 번에 업데이트 하는 메서드 총 2개로 작성하려 했으나,
유연성을 고려하여 각각의 필드를 업데이트 할 수 있는 메서드를 생성했습니다.

xxbeann committed:
refactor: 필드 검증 메서드 변경

xxbeann committed:
feat: 프로필 업데이트 컨트롤러 작성

xxbeann committed:
feat: 사용자 프로필 업데이트 서비스 로직 작성

xxbeann committed:
feat: 유효성 검증 추가

xxbeann committed:
feat:이메일 필드 NotBlank 검증 추가

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/user/domain/User.java:
```diff
@@ -35,6 +35,8 @@ public class User extends BaseEntity {
     @Column(unique = true)
     private String nickname;
 
+    private String etc;
+
```
기타 정보가 신체에 대한 정보라서 physicalInfo에 넣으면 어떨까요?

xxbeann replied:
반영하겠습니다!

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/user/domain/vo/UserInfo.java:
```diff
@@ -30,18 +30,21 @@ public UserInfo(String email, String name){
-    public void updateBasicInformation(String email, String phoneNumber){
+    public void updateEmail(String email){
         validateEmail(email);
-        validatePhoneNumber(phoneNumber);
         this.email = email;
-        this.phoneNumber = phoneNumber;
     }
 
     public void updateName(String name){
         validateName(name);
         this.name = name;
     }
 
+    public void updatePhoneNumber(String phoneNumber){
+        validatePhoneNumber(phoneNumber);
+        this.phoneNumber = phoneNumber;
+    }
```
`updateBasicInformation()` 메소드 내에서 email과 name을 업데이트 하는 방식이 아닌 각각 별도의 메소드로 쪼갠 이유가 있으실까요?

xxbeann replied:
그러게요....?
시간이 조금 지나서 기억이 가물가물하지만, 메서드 명이 아마 적합하지 않다고 판단했던 것 같습니다.
User 엔터티에 있는 `registerBasicInformation` 메서드를 보면, 기본정보(닉네임, 전화번호, 이메일)을 등록하는 역할을 합니다.
여기서 주목한 건 기본정보가 닉네임, 전화번호, 이메일을 포함한다는 것이었습니다.
그리고 userInfo에 있는 `updateBasicInformation`을 호출하는데, 이때 이메일과 전화번호만 업데이트 합니다.
이런 동작이 `updateBasicInformation`과는 어울리지 않다고 생각했습니다.
(왜냐하면, 해당 메서드를 보면 기본정보 3가지를 모두 업데이트 해야할 것 같은데 실제로는 2가지만 업데이트 하기 때문)
그래서 메서드를 분리시켰던 것 같습니다...!

Woojae-Jeong reviewed (commented):
>physicalInfo 내에서 검증 로직이 필요한가? 

넵 필요합니다. 이전 pr 리뷰에서 null값 검사 시 던지는 에러로 자바 예외가 적합하다고 제 의견을 말씀드렸을 때, 보충 설명으로 적었던 동료나 내가 해당 객체를 어디서 생성할 줄 모른다는 걸 기억하실겁니다. 그래서 physicalInfo가 프로필 생성에 뿐만 아니라 다른 곳에서도 생성이 될 수 있기에 `vo 객체들은 자체적으로 null값 검사와 같이 비정상적인 데이터인지에 대한 검사가 필요합니다.`

더불어 그럼 도메인 모델인 엔티티에서도 필요하지 않냐? 라고 의문이 드실겁니다. 도메인 모델에서의 검증은 비정상적인 데이터에 대한 검증이 아닌 `비즈니스 규칙에 대한 검증이 이루어져야 합니다.` 예를 들어서 `닉네임은 유저의 나이, 이름이 들어가야 한다` 라는 비즈니스 규칙이 있다면, 이는 단순 비정상적인 데이터 검증이 아닌 비즈니스 규칙 검증이 이루어져야 하므로 이를 도메인 모델에서 수행하게 됩니다

아래는 참고글이며 읽어보시면 잘 이해가 될 것 같네요

https://jaeseo0519.tistory.com/421

xxbeann committed:
feat: updateUserProfile메서에서 physicalInfo NULL 방어로직 작성

회원가입시 physicalInfo객체는 생성되지 않아 null상태입니다.
그 상태에서 updateUserProfile메서드를 호출하게되면 NPE이 발생합니다.
해당 메서드가 다른 곳에서도 사용될 수 있기에 범용성을 고려하여,
Null 가드를 추가했습니다.
해당부분 별도 메서드로 분리가 필요할 것 같다고 생각되면 리뷰 부탁드립니다.

xxbeann committed:
feat: physicalInfo null 예외처리

사용자 프로필을 조회해올때, physicalInfo가 입력되었다고 가정된 상황만 고려하였습니다.
그래서 null 예외처리를 할 필요가 없다고 생각했습니다.
다만, 어디 화면에서 프로필조회를 할 지 모를 뿐더러, 실제로 전화번호만 가져와야 하는 기능이 필요해,
NPE이 발생한 경우가 있었습니다. 따로 api를 만드는 것은 비효율적이며, 해당 API의 범용적인 사용을 고려하여
NULl 방오 로직을 추가하였습니다.

xxbeann commented:
해당 블로그 저자 상당히 마음에 드네요..!
자세한 답변 감사합니다! 그리고 늦어서 죄송합니다!
리팩토링 해보겠습니다

xxbeann committed:
refactor: user etc 필드 physicalInfo로 변경

xxbeann committed:
feat: VO 검증 로직 추가

vo 검증로직을 추가했습니다.
해당 vo는 회원가입시 생성되지 않습니다. 즉 회원가입시점부터 null 객체이기때문에
nullable = false는 적용하지 않았습니다. npe이 발생할 수 있는 부분은 적절히 null 예외로직을 적용했습니다.
reqDTOD와 똑같이 weight, height, etc에 대해 최대 길이(범위)검증로직을 추가했습니다.
bloodType은 ENUM이 타입 보장을 해주기 때문에 검증로직을 추가하지 않았습니다.

xxbeann committed:
chore: weight, height 최대값 범위 조절

xxbeann committed:
feat: validateRange에 NPE 예외처리

xxbeann committed:
chore: confilct resolve
